test/jaxp/javax/xml/jaxp/unittest/xpath/XPathChildTest.java

```diff
@@ -0,0 +1,217 @@
+                {"/store/child::book[last()]/author", AUTHOR_2},
+
+                // descendant
+                {"//book/child::*[2]", AUTHOR_1},
+                {"//child::*[child::author]/author", AUTHOR_1},
+                {"//child::*[child::author][2]/author", AUTHOR_2},
+                {"//child::node()/child::author", AUTHOR_1},
+                {"//child::node()[child::author]/author", AUTHOR_1},
+                {"//child::node()[child::author][2]/author", AUTHOR_2},
+
+                // parent node
+                {"//child::book/../child::book/child::author", AUTHOR_1},
+
+                // dot reference
+                {"//child::book/./child::author", AUTHOR_1},
+                {"//child::node()/./child::author", AUTHOR_1},
+                {"//././/./child::author", AUTHOR_1},
+
+                // attributes
+                {"/store/child::book[@id=1]/author", AUTHOR_1},
+                {"/store/child::book[attribute::id=1]/author", AUTHOR_1},
+                {"/store/child::book[@id]/author", AUTHOR_1},
+                {"/store/child::book[@id=1][@lang='en']/author", AUTHOR_1},
+                {"/store/child::book[@lang='en'][1]/author", AUTHOR_1},
+                {"/store/child::book[child::isbn='1234']/author", AUTHOR_1},
+                {"/store/child::book[@lang='en' and " +
+                        "child::isbn='1234']/author", AUTHOR_1},
+                {"/store/child::*[@lang='en'][2]/author", AUTHOR_2},
+                {"/store/child::node()[@id='1']/author", AUTHOR_1},
+                {"/store/child::node()[@lang='en'][2]/author", AUTHOR_2},
+                {"/store/child::*[child::author][child::title][@id='2']/author",
+                        AUTHOR_2},
+                {"/store/child::*[child::author or child::ssn][@id='2']/author",
+                        AUTHOR_2},
+                {"/store/child::*[child::*]/author", AUTHOR_1},
+                {"/store/child::*[attribute::*]/author", AUTHOR_1},
+                {"/store/*[*][*][*][*][*][*][*][*]/author", AUTHOR_1},
+                {"/store/*[@*][@*][@*][@*][@*][@*][@*][@*]/author", AUTHOR_1},
+                {"//author[@*]", AUTHOR_1},
+
+                // text node
+                {"/store/book[1]/isbn/child::text()/../../author", AUTHOR_1},
+                {"/store/book/isbn[child::text()='5678']/../author", AUTHOR_2},
+                {"/store/book/isbn[.='5678']/../author", AUTHOR_2},
+
+                // count child nodes
+                {"/store/book[count(./child::author)]/author", AUTHOR_1},
+                {"/store/book[count(child::author)]/author", AUTHOR_1},
+                {"/store/book[count(../child::book)]/author", AUTHOR_2},
+        };
+    }
+
+    /*
+     * DataProvider: provides invalid XPath expression and expected exception
+     */
+    @DataProvider(name = "invalidExp")
+    public Object[][] getInvalidExp() {
+        return new Object[][]{
+                // NullPointerException
```

Comment:
These NPE tests need to be revisited. First of all, by the spec, NPE shall throw only "If expression or returnType is null" (and that's already covered in the Exception test). Secondly, these NPEs were not thrown by the eval process, it was merely because the NodeList was empty or Node was null. The expressions themselves were not "invalid".

Reply:
Agreed. These expressions are not invalid and they are more on a focus of zero children rather than NPE. I can move them to a different test.